Add observability in the cloud post (#11)
bespinian/blog: content/posts/observability-in-the-cloud.md

---
title: Observability in the Cloud
author: Gabriel Koch
comments: true
date: 2022-04-01
---

This post aims to explain how observability and monitoring should be distinguished, what the problems with "traditional" monitoring is when we deal with distributed cloud systems and how observability can alleviate these problems.

> Finally I roughly outline how it should be approached.

Reviewer: ```suggestion
Finally I roughly outline how observability should be approached.
```

## Observability vs. Monitoring

> For the difference between monitoring and observability, I recommend Cindy Sridharan's book Distributed Systems Observability and Charity Major's various blog posts on the topic (e.g. [Observability is a Many-Splendored Definition](https://charity.wtf/2020/03/03/observability-is-a-many-splendored-thing/)) who go into more detail.

Reviewer: ```suggestion
For detailed accounts of the difference between monitoring and observability, I recommend Cindy Sridharan's book Distributed Systems Observability and Charity Major's various blog posts on the topic (e.g. [Observability is a Many-Splendored Definition](https://charity.wtf/2020/03/03/observability-is-a-many-splendored-thing/)).
```

> Here, I summarize the main points and draw some conclusions.

Reviewer: ```suggestion
Here, I summarize the main points of these accounts and draw some conclusions.
```

Monitoring is what most of us are used to.
It encompasses collection of telemetry (metrics, logs and traces), visualization through dashboards and alerting, with metrics being the primary data type used.
Monitoring is primarily the process of collecting telemetry and acting upon it.

> It deals with the known-unknowns (e.g. we know that we may run out of memory or disk and can monitor it, but we do not precisely know when and because of which processes it happens).

Reviewer: ```suggestion
It deals with the known-unknowns (e.g. we know that we may run out of memory or disk and we can monitor it, but we do not precisely know when it will happen and which processes causes it to happen).
```

Observability, on the other side, describes a property of a system.
A system is observable, if you can deduce its internal state (such as the value of its internal variables and position in the callstack), from the outputs you can observe.

> These outputs can be specific behaviours exhibited to users, and in practice rely heavily on telemetry we collect from the system's environment (such as the machine it's running on) or deliberately output from the system.

Reviewer: ```suggestion
These outputs can be specific behaviours exhibited to users, and in practice rely heavily on telemetry which we collect from the system's environment (such as the machine it is running on) or which we deliberately emit from the system.
```

> A system can be described as "more observable", when we can more accurately describe its internal state, and fully observable, when we can always describe its internal state with 100% accuracy.

Reviewer: ```suggestion
A system can be described as "more observable", when we can more accurately describe its internal state, and "fully observable", when we can always describe its internal state with 100% accuracy.
```

> Making a system observable usually requires using monitoring tools and processes, but it allows us to deal with a larger set of problems, namely, the unkown-unknowns, and effectively troubleshoot issues that we could not really imagine before they arose.

Reviewer: ```suggestion
Making a system observable usually requires using monitoring tools and processes, but it enables us to deal with a larger set of problems, namely, the unkown-unknowns, and to effectively troubleshoot issues that we could not really imagine before they arose.
```

## The Limits of Monitoring

So, now that we know the difference between monitoring and observability, how did monitoring help us deal with problems in the past?

> Once something goes wrong, we may use the alerts and dashboards to get an overview and over which application, machine, or component is affected.

Reviewer: ```suggestion
Once something goes wrong, we may use the alerts and dashboards to get an overview over which application, machine, or component is affected.
```

Additionally, we head into the logs, to try and figure out where exactly things went wrong.

> If it is a known issue, we can usually apply a known fix.

Reviewer: ```suggestion
If the root cause is a known issue, we can usually apply a known fix.
```

> But if it's an unknown issue, we often have to dig through extensive logs, manually correlate them and finally we may start up the application with a debugger, and try to reproduce the exact conditions.

Reviewer: ```suggestion
But if the root cause is an unknown issue, we often have to dig through extensive logs, manually correlate them and finally we may have to start up the application with a debugger in order to try to reproduce the exact conditions.
```

We iteratively go back to the logs, to grab more pieces of evidence and approximate the state of the system when the error occurred.

> Telemetry in our monitoring was usually highly aggregated already, but this is a minor issue, since most problems are clearly attributable to a specific system and we can effectively use a debugger to investigate it.

Reviewer: The point of this sentence is not entirely clear to me. Could this be rephrased somehow?

Author: I rephrased it slightly as well as the section after it.

However, when we develop applications for the clouds or when we refactor monolithic applications, we often strive to implement a microservice architecture.
This means, we have a lot more different runtimes handling work and interacting via the network.

Additionally, the following practices are being adopted more widely:

- dynamic scaling and disposable application instances
- asynchronous processing, by using queues and streaming services
- blue/green, canary or rolling deployments
- feature flags
- shorter deployment cycles

> These changes lead to a more complex application landscape, which makes debugging more and more infeasible.

Reviewer: ```suggestion
These changes lead to a more complex application landscape, which makes debugging increasingly difficult.
```

Instead of running one service in a debugger and recreating its state, we may need to start various services and try to reproduce state in each of them.
Our telemetry may also become less expressive and we may have to ask ourselves additional questions, such as:

- has a deployment just happened?
- what versions are actually deployed?
- which versions (if multiple are running) show anomalous behaviour
- which feature flags were enabled for the requests/tasks that failed?

> To ensure, we still get some meaningful insights from our telemetry, we need to annotate it with additional information such as build ID, hostname, IP address, enabled feature flags and more.

Reviewer: ```suggestion
To ensure, we still get some meaningful insights from our telemetry, we need to annotate its data with additional information such as build ID, hostname, IP address, enabled feature flags and more.
```

Author: This change sounds very odd to me

> It also brings about new problems: adding these dimensions to our telemetry makes the storage and querying systems much slower and/or more expensive.

Reviewer: ```suggestion
This data enrichment brings about new problems: adding more dimensions to our telemetry data makes the storage and querying systems much slower and/or more expensive.
```

At the same time, we still need to make a number of queries to correlate information from multiple source systems (microservices) to slowly piece together enough context to guess what happened.

## Closing the Observability Gap in Distributed Cloud Systems

Unfortunately, this gap cannot be closed by the simple addition of a new tool or library.
Instead, there are two primary challenges that should be tackled by site reliability engineers who aim to improve observability in their systems.

> First, each individual service should be instrumented to keep track of its context and emit arbitrarily wide events containing all relevant context when a request/task processing finishes (successfully or with an exception).

Reviewer: ```suggestion
First, each individual service should be instrumented to keep track of its context. Furthermore, each service should emit arbitrarily wide events containing the relevant context when a request/task processing finishes (successfully or with an exception).
```

To do this, whenever a workload is started (e.g. a request enters a system), an event context object should be initialized and prepopulated with all relevant data.
As the workload is being processed, this object should be enriched and whenever another service is called, relevant context should be passed downstream.
When the request finishes, the whole context event should be passed to the central storage system.

> Many logging libraries do support such use cases, but it often depends on the framework used.

Reviewer: ```suggestion
Many logging libraries support such use cases, but they also need to be implemented correctly in the frameworks used.
```

Ensuring, that these events are properly enriched is the responsibility of developers and site reliability engineers.
The goal should be, that problems can usually be pinpointed by looking at a single event, without having to painstakingly correlate dozens of events or reproduce an issue through incremental trial and error.

> As a second challenge, an event storage and querying system must be put in place, that is able to reasonably process these wide events and allows querying and searching.

Reviewer: ```suggestion
As a second challenge, an event storage and querying system must be put in place, which is able to process these wide events reasonably and which allows querying and searching.
```

> Storing arbitrarily wide events is crucial and querying across all dimensions is crucial to avoid limitations to the sort of questions you can ask, but it is not supported by many monitoring systems, especially since it's usually very resource intensive.

Reviewer: ```suggestion
Storing arbitrarily wide events and querying across all dimensions is crucial in order to avoid limitations to the sort of questions you can ask. However, it is not supported by many monitoring systems, especially since it is usually very resource intensive.
```

Some observability / monitoring tools, such as Elasticsearch, already support such use cases.

> These systems may get expensive and require fine-tuning and optimization in itself to ensure they perform well while staying cost-effective.

Reviewer: ```suggestion
These systems may get expensive and require fine-tuning and optimization in their own right to ensure that they perform well while staying cost-effective.
```

## Summary

> While the term monitoring primarily describes the process of collecting telemetry and defining some automated responses, the term observability goes wider and describes the property of a system of how well you can infer its internal state from its outputs.

Reviewer: ```suggestion
While the term monitoring primarily describes the process of collecting telemetry and defining some automated responses, the term observability goes further. It describes the property of a system, namely how well you can infer the system's internal state from its outputs.
```

Monitoring is primarily focused on diagnosing problems that occur in similar ways as problems we saw before, while true observability allows you to effectively diagnose any problem, even completely novel ones.

> The crucial actions any site reliability engineer should take is to start collecting all relevant context information while a workload is processed, and emit it in an event at the end, as well as ensuring a system for collecting and effectively querying these events is in place.

Reviewer: ```suggestion
The crucial actions which any site reliability engineer should take are 

1. to start collecting all relevant context information while a workload is processed
1. to emit this information in an event at the end 
1. to ensure that a system for collecting and effectively querying these events is in place.
```

---
title: Observability in the Cloud
author: Gabriel Koch
comments: true
date: 2022-04-01
---

> This post aims to explain how observability and monitoring should be distinguished, what the problems with "traditional" monitoring is when we deal with distributed cloud systems and how observability can alleviate these problems.

Reviewer 2: ```suggestion
This blog post aims to explain how observability and monitoring should be distinguished, what the problems with "traditional" monitoring are when we deal with distributed cloud systems and how observability can alleviate these problems.
```

> Finally I roughly outline how observability should be approached.

Reviewer 2: ```suggestion
Finally, it roughly outlines how observability should be approached.
```

## Observability vs. Monitoring

For detailed accounts of the difference between monitoring and observability, I recommend Cindy Sridharan's book Distributed Systems Observability and Charity Major's various blog posts on the topic (e.g. [Observability is a Many-Splendored Definition](https://charity.wtf/2020/03/03/observability-is-a-many-splendored-thing/)).

Here, I summarize the main points of these accounts and draw some conclusions.

Monitoring is what most of us are used to.

> It encompasses collection of telemetry (metrics, logs and traces), visualization through dashboards and alerting, with metrics being the primary data type used.

Reviewer 2: ```suggestion
It encompasses the collection of telemetry (metrics, logs and traces) and its visualization through dashboards and alerting, with metrics being the primary data type used.
```

Monitoring is primarily the process of collecting telemetry and acting upon it.

> It deals with the known-unknowns (e.g. we know that we may run out of memory or disk and we can monitor it, but we do not precisely know when it will happen and which processes causes it to happen).

Reviewer 2: ```suggestion
It deals with the known-unknowns (e.g. we know that we may run out of memory or disk space and we can monitor it, but we do not precisely know when it will happen and which processes cause it to happen).
```

Observability, on the other side, describes a property of a system.
A system is observable, if you can deduce its internal state (such as the value of its internal variables and position in the callstack), from the outputs you can observe.

> These outputs can be specific behaviours exhibited to users, and in practice rely heavily on telemetry which we collect from the system's environment (such as the machine it is running on) or which we deliberately emit from the system.

Reviewer 2: ```suggestion
These outputs can be specific behaviors exhibited to users, and in practice rely heavily on telemetry which we collect from the system's environment (such as the machine it is running on) or which we deliberately emit from the system.
```

Our entire blog uses AE :wink:

In a microservice architecture, we have a lot more different runtimes handling work and interacting via the network.
Additionally, the following practices are being adopted more widely, when we design applications for clouds:

- dynamic scaling and disposable application instances
- asynchronous processing, by using queues and streaming services
- blue/green, canary or rolling deployments
- feature flags
- shorter deployment cycles

These changes lead to a more complex application landscape, which makes debugging increasingly difficult.
Instead of running one service in a debugger and recreating its state, we may need to start various services and try to reproduce state in each of them.
Our telemetry may also become less expressive and we may have to ask ourselves additional questions, such as:

- has a deployment just happened?
- what versions are actually deployed?
- which versions (if multiple are running) show anomalous behaviour
- which feature flags were enabled for the requests/tasks that failed?

> To ensure, we still get some meaningful insights from our telemetry, we need to annotate it with additional information such as build ID, hostname, IP address, enabled feature flags and more.

Reviewer 2: ```suggestion
To ensure that we still get meaningful insights from our telemetry, we need to annotate it with additional information such as build ID, hostname, IP address, enabled feature flags and more.
```

This data enrichment brings about new problems: adding more dimensions to our telemetry data makes the storage and querying systems much slower and/or more expensive.
At the same time, we still need to make a number of queries to correlate information from multiple source systems (microservices) to slowly piece together enough context to guess what happened.

## Closing the Observability Gap in Distributed Cloud Systems

Unfortunately, this gap cannot be closed by the simple addition of a new tool or library.

> Instead, there are two primary challenges that should be tackled by site reliability engineers who aim to improve observability in their systems.

Reviewer 2: ```suggestion
Instead, there are two primary challenges that should be tackled by site reliability engineers who aim to improve the observability in their systems.
```

First, each individual service should be instrumented to keep track of its context. Furthermore, each service should emit arbitrarily wide events containing the relevant context when a request/task processing finishes (successfully or with an exception).

> To do this, whenever a workload is started (e.g. a request enters a system), an event context object should be initialized and prepopulated with all relevant data.

Reviewer 2: ```suggestion
To do this, whenever a workload is processed (e.g. a request enters a system), an event context object should be initialized and prepopulated with all relevant data.
```

The start of a workload, in my intuition, would rather refer to the startup of an application instead of it processing a request.
